### PR TITLE
KnownBits: refine high-bits of mul in signed case

### PR DESCRIPTION
**Pull request by artagnon**
KnownBits::mul suffers from the deficiency that it doesn't account for signed inputs. Fix it by refining known leading zeros when both inputs are signed, and setting known leading ones when one of the inputs is signed. The strategy we've used is to still use umul_ov, after adjusting for signed inputs, and setting known leading ones from the negation of the result, when it is known to be negative, noting that a possibly-zero result is a special case.

-- 8< --
Based on #113050.

### CODE REVIEW

**Comment by goldsteinn** [comment]
How does this compare to Jay's impl https://github.com/llvm/llvm-project/issues/86671#issuecomment-2023064384?

**Comment by artagnon** [comment]
> How does this compare to Jay's impl [#86671 (comment)](https://github.com/llvm/llvm-project/issues/86671#issuecomment-2023064384)?

My patch makes KnownBits::mul strictly better for the high bits: I don't touch the low bits. Jay's implementation attempts to rewrite mul completely, at the cost of compile time.

**Comment by goldsteinn** [comment]
> > How does this compare to Jay's impl [#86671 (comment)](https://github.com/llvm/llvm-project/issues/86671#issuecomment-2023064384)?
> 
> My patch makes KnownBits::mul strictly better for the high bits: I don't touch the low bits. Jay's implementation attempts to rewrite mul completely, at the cost of compile time.

Strictly better than current or jays impl?

**Review by jayfoad — COMMENTED** [review]
The implementation looks fine to me. As you say it's strictly better than the status quo.

I'd prefer to put the ad hoc tests in `unittests/Support/KnownBitsTest.cpp` since it's a more direct way of testing the KnownBits implementation than going via ValueTracking.

For future work: I suspect that what you have implemented here is not quite optimal for the high bits. We could fix that and extend #113316 to check optimality of high bits as well as low bits.

**Comment by artagnon** [comment]
> I'd prefer to put the ad hoc tests in `unittests/Support/KnownBitsTest.cpp` since it's a more direct way of testing the KnownBits implementation than going via ValueTracking.

I'm not sure we'd want to pollute the unit tests with ad-hoc tests: to me, KnownBitsTest.cpp is a collection of disciplined tests that should never fail. We already have several knownbits-*.ll in ValueTracking/test, and I think adding knownbits-mul.ll works in practice, as long as we pre-commit a reduced test, show non-reduction, and show a reduction after the patch.
 
> For future work: I suspect that what you have implemented here is not quite optimal for the high bits. We could fix that and extend #113316 to check optimality of high bits as well as low bits.

I also suspect that it's not optimal for high bits: I think we need some logic similar to the low-bits logic later in the function. Will think about a follow-up when I have some free time, if you don't beat me to it first.

**Comment by artagnon** [comment]
> I'm generally in favor of this, but I'm also in favors of jays impl getting in unless the compile time data is really bad.

I think looping through bits is unnecessary, as low-bits are already optimal, and we need something like that for high bits: we can work on a follow-up.

**Comment by artagnon** [comment]
> > For future work: I suspect that what you have implemented here is not quite optimal for the high bits. We could fix that and extend #113316 to check optimality of high bits as well as low bits.
> 
> I also suspect that it's not optimal for high bits: I think we need some logic similar to the low-bits logic later in the function.

Quick remark: I think this is sub-optimal due to the mul_ov overflowing in many cases.

**Comment by jayfoad** [comment]
> > I'd prefer to put the ad hoc tests in `unittests/Support/KnownBitsTest.cpp` since it's a more direct way of testing the KnownBits implementation than going via ValueTracking.
> 
> I'm not sure we'd want to pollute the unit tests with ad-hoc tests: to me, KnownBitsTest.cpp is a collection of disciplined tests that should never fail.

There's no question of "pollution". It's the correct place for KnownBits tests. Yes we have tried to make most of the existing tests "disciplined" and exhaustive, but ad hoc tests can go in there too. They also have the advantage that it's much more obvious what cases are being tested - you don't have to craft IR with `and`s and `or`s to persuade ValueTracking to create the KnownBits objects you want to test.

**Comment by jayfoad** [comment]
> > > For future work: I suspect that what you have implemented here is not quite optimal for the high bits. We could fix that and extend #113316 to check optimality of high bits as well as low bits.
> > 
> > 
> > I also suspect that it's not optimal for high bits: I think we need some logic similar to the low-bits logic later in the function.
> 
> Quick remark: I think this is sub-optimal due to the mul_ov overflowing in many cases.

Yeah, I have thought about this a bit more. I suspect that making high bits optimal in all cases is just as hard as making "middle" bits optimal. So we might have to give up on that goal.

**Comment by artagnon** [comment]
> > > > For future work: I suspect that what you have implemented here is not quite optimal for the high bits. We could fix that and extend #113316 to check optimality of high bits as well as low bits.
> > > 
> > > 
> > > I also suspect that it's not optimal for high bits: I think we need some logic similar to the low-bits logic later in the function.
> > 
> > 
> > Quick remark: I think this is sub-optimal due to the mul_ov overflowing in many cases.
> 
> Yeah, I have thought about this a bit more. I suspect that making high bits optimal in all cases is just as hard as making "middle" bits optimal. So we might have to give up on that goal.

I have a question. Can we simply mul_ov with twice the bitwidth to get the optimal result?

**Comment by jayfoad** [comment]
If you multiply with twice the bit width, it will never overflow.

**Comment by artagnon** [comment]
> > > I'd prefer to put the ad hoc tests in `unittests/Support/KnownBitsTest.cpp` since it's a more direct way of testing the KnownBits implementation than going via ValueTracking.
> > 
> > 
> > I'm not sure we'd want to pollute the unit tests with ad-hoc tests: to me, KnownBitsTest.cpp is a collection of disciplined tests that should never fail.
> 
> There's no question of "pollution". It's the correct place for KnownBits tests. Yes we have tried to make most of the existing tests "disciplined" and exhaustive, but ad hoc tests can go in there too. They also have the advantage that it's much more obvious what cases are being tested - you don't have to craft IR with `and`s and `or`s to persuade ValueTracking to create the KnownBits objects you want to test.

Did you mean something like this? If you can check that it is okay, I will push it.

```cpp

TEST(KnownBitsTest, MulHighBits) {
  unsigned Bits = 8;
  SmallVector<std::pair<int, int>, 4> TestPairs = {
      {2, 4}, {-2, -4}, {2, -4}, {-2, 4}};
  for (auto [K1, K2] : TestPairs) {
    KnownBits Known1(Bits), Known2(Bits);
    if (K1 > 0) {
      Known1.Zero |= ~(K1 | 1);
      Known1.One |= 1;
    } else {
      Known1.One |= K1;
    }
    if (K2 > 0) {
      Known2.Zero |= ~(K2 | 1);
      Known2.One |= 1;
    } else {
      Known2.One |= K2;
    }
    KnownBits Computed = KnownBits::mul(Known1, Known2);
    KnownBits Exact(Bits);
    Exact.Zero.setAllBits();
    Exact.One.setAllBits();

    ForeachNumInKnownBits(Known1, [&](const APInt &N1) {
      ForeachNumInKnownBits(Known2, [&](const APInt &N2) {
        APInt Res = N1 * N2;
        Exact.One &= Res;
        Exact.Zero &= ~Res;
      });
    });

    APInt Mask = APInt::getHighBitsSet(
        Bits, (Exact.Zero | Exact.One).countLeadingOnes());
    Exact.Zero &= Mask;
    Exact.One &= Mask;
    Computed.Zero &= Mask;
    Computed.One &= Mask;
    EXPECT_TRUE(checkResult("mul", Exact, Computed, {Known1, Known2},
                            /*CheckOptimality=*/true));
  }
}
```

**Comment by jayfoad** [comment]
Sure that looks good, with a couple of comments to explain what is going on. It is much more general than I was expecting.

**Comment by artagnon** [comment]
> If you multiply with twice the bit width, it will never overflow.

Right. I _think_ I can work on a follow-up doing the mul_ov with twice the bitwidth, and make the high bits exhaustive. Will investigate after this patch lands.

File: llvm/lib/Support/KnownBits.cpp [hunk 1]
```diff
@@ -796,19 +796,26 @@ KnownBits KnownBits::mul(const KnownBits &LHS, const KnownBits &RHS,
...
-  unsigned LeadZ = HasOverflow ? 0 : UMaxResult.countl_zero();
+  APInt Result = MaxLHS.umul_ov(MaxRHS, HasOverflow);
+  bool NegResult = LHS.isNegative() ^ RHS.isNegative();
+  unsigned LeadZ = 0, LeadO = 0;
+  if (!HasOverflow) {
+    // Do not set leading ones unless the result is known to be non-zero.
+    if (NegResult && LHS.isNonZero() && RHS.isNonZero())
+      LeadO = (-Result).countLeadingOnes();
+    else if (!NegResult)
+      LeadZ = Result.countLeadingZeros();
```
Comment (goldsteinn) [inline llvm/lib/Support/KnownBits.cpp:817]: Nit: Think slightly more concise to do:
```
if(LHS.isNegative() ^ RHS.isNegative()) 
  // Neg Case
else if (LHS.isNonZero() && RHS.isNonZero())
 // Non-Neg Result.
```

**Comment by artagnon** [comment]
> Right. I _think_ I can work on a follow-up doing the mul_ov with twice the bitwidth, and make the high bits exhaustive. Will investigate after this patch lands.

I thought about it some more, and have come to the same conclusion as you: making high bits optimal is as hard as making the entire mul optimal. I was hoping to rely on remainders, but urem isn't optimal either.

**Comment by artagnon** [comment]
Gentle ping. An alternative is to directly review #114211.

**Comment by nikic** [comment]
For the future, can you please not create separate PRs for test additions and instead include them as a separate commit? Otherwise we cannot test the impact on llvm-opt-benchmark.

**Comment by artagnon** [comment]
> For the future, can you please not create separate PRs for test additions and instead include them as a separate commit? Otherwise we cannot test the impact on llvm-opt-benchmark.

The test addition is included as the first commit in this PR, no?

**Comment by artagnon** [comment]
> And I see now that your follow-up PR was already tested in [dtcxzyw/llvm-opt-benchmark#1591](https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1591), and does not seem to provide benefit on real code.

True, and that came as a very unfortunate surprise, as I spent quite a lot of time on these PRs :(

File: llvm/lib/Support/KnownBits.cpp [hunk 2]
```diff
@@ -796,19 +796,25 @@ KnownBits KnownBits::mul(const KnownBits &LHS, const KnownBits &RHS,
...
-  // Compute the high known-0 bits by multiplying the unsigned max of each side.
-  // Conservatively, M active bits * N active bits results in M + N bits in the
-  // result. But if we know a value is a power-of-2 for example, then this
-  // computes one more leading zero.
-  // TODO: This could be generalized to number of sign bits (negative numbers).
-  APInt UMaxLHS = LHS.getMaxValue();
-  APInt UMaxRHS = RHS.getMaxValue();
-
-  // For leading zeros in the result to be valid, the unsigned max product must
+  // Compute the high known-0 or known-1 bits by multiplying the max of each
```
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:799]: I think you can wrap the whole of this section in `if (!LHS.isSignUnknown() && !RHS.isSignUnknown)`. If either sign was unknown then we would not get any useful high-zeros or high-ones info from this calculation. That would resolve my confusion about your use of `isNegative` below by making it clear that there is no "unknown sign" case to worry about - both operand are known negative or known positive.

File: llvm/lib/Support/KnownBits.cpp [hunk 3]
```diff
@@ -796,19 +796,25 @@ KnownBits KnownBits::mul(const KnownBits &LHS, const KnownBits &RHS,
...
+  // side. Conservatively, M active bits * N active bits results in M + N bits
+  // in the result. But if we know a value is a power-of-2 for example, then
+  // this computes one more leading zero or one.
+  APInt MaxLHS = LHS.isNegative() ? LHS.getMinValue().abs() : LHS.getMaxValue(),
+        MaxRHS = RHS.isNegative() ? RHS.getMinValue().abs() : RHS.getMaxValue();
```
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:804]: Can you remove the `.abs()` calls here and use `smul_ov` below instead?

**Comment by artagnon** [comment]
@jayfoad Thanks for the review. However, after some reflection and offline discussion with @nikic, we have decided to close both PRs, since the patches don't have any benefit on real-world code, and just add compile-time. A completely optimal mul might be beneficial from a testing point of view but, as the real-world benchmarks indicate, a purely academic exercise.

**Pull request closed without merging**